codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java

```diff
@@ -150,6 +160,33 @@ protected final boolean validateHeaders() {
      * @return a new {@link Http2Headers} object which will store the results of the decode operation.
      */
     protected Http2Headers newHeaders() {
-        return new DefaultHttp2Headers(validateHeaders, (int) headerArraySizeAccumulator);
+        return new DefaultHttp2Headers(validateHeaders, (int) headerArraySizeAccumulator, true);
```

Comment:
note that headers are modifiable by default. So after the headers are decoded they maybe modified and conceivably add another pseudo header (e.g. a proxy use case). I'm hesitant to generate a `Http2Headers` objects which does not enforce the iteration order required by the specification under these constraints. IMHO it would be safer just to isolate this behavior to the unit tests if that is where it is required.

Reply:
@Scottmitch @madgnome +1

Reply:
Oh damn, excellent point.

> IMHO it would be safer just to isolate this behavior to the unit tests if that is where it is required.

We need to know the insertion order to be able to detect if the headers are out of order when we decode them. 

Back to the drawing board...